### PR TITLE
Add mechanism for 'complaining' about autocall / autokey

### PR DESCRIPTION
**Pull request by asottile**

### CODE REVIEW

File: Cheetah/c/_namemapper.c [hunk 1]
```diff
@@ -231,6 +263,19 @@ static PyObject *PyNamemapper_valueForName(PyObject *obj, char *nameChunks[], in
 /* Now the wrapper functions to export into the Python module */
 /* *************************************************************************** */
 
+static PyObject* namemapper_set_complain_func(PYARGS) {
```
Comment (bukzor) [inline Cheetah/c/_namemapper.c]: a link to your source material would be useful https://docs.python.org/2/extending/extending.html#calling-python-functions-from-c

File: Cheetah/c/_namemapper.c [hunk 2]
```diff
@@ -255,7 +300,8 @@ static PyObject *namemapper_valueForName(PYARGS)
 
     createNameCopyAndChunks();
 
-    theValue = PyNamemapper_valueForName(obj, nameChunks, numChunks, executeCallables, useDottedNotation);
+    /* passing 0 for isFirst, VFN is only called like VFN(VFFSL(...), ...) */
```
Comment (bukzor) [inline Cheetah/c/_namemapper.c]: I don't think that's a valid assumption.
People can easily call VFN however they like. It's part of the documented interface.

Comment (asottile) [inline Cheetah/c/_namemapper.c]: `VFN` / `VFSL` / `VFFSL` are implementation details, nowhere are they documented as public functions.

File: tests/complain_test.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+@with_raising_complain
+def test_does_not_raise_when_key_and_attr_are_available():
+    class SneakyDictAttrThing(object):
+        def __init__(self, **kwargs):
+            self.__dict__.update(kwargs)
+
+        def __getitem__(self, val):  # pragma: no cover
+            return getattr(self, val)
+
+    compile_to_class(
```
Comment (bukzor) [inline tests/complain_test.py]: assert something please. 

`== 'baz'`, I think.

File: tests/complain_test.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        'Autokey detected while processing name `bar`',
+    )
+
+
+def test_callable_after_raising_complain():
+    with autocall_autokey_complain(raise_complaint):
+        pass
+
+    # Should not raise
+    compile_to_class(
```
Comment (bukzor) [inline tests/complain_test.py]: assert something please

File: tests/complain_test.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    with pytest.raises(ComplaintError) as excinfo:
+        inst.respond()
+    assert excinfo.value.args == (
+        'Autocall detected while processing name `foo`',
+    )
+
+
+@with_raising_complain
+def test_does_not_raise_on_locals():
+    compile_to_class(
```
Comment (bukzor) [inline tests/complain_test.py]: assert "bar"

File: Cheetah/c/_namemapper.c (hunk 2 above)
Comment (bukzor) [inline Cheetah/c/_namemapper.c]: More importantly, if anyone uses this stuff, they will get strictly too many errors.

**Comment by bukzor** [comment]
:shipit:

**Pull request merged**